File: do-not-share/jads-android/app/src/main/kotlin/com/jads/ui/livetrack/LiveTrackViewModel.kt
// T06 — ViewModel for live drone tracking on remote viewer
package com.jads.ui.livetrack

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jads.websocket.JadsWebSocketClient
import com.jads.websocket.WsEvent
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

data class DronePosition(
    val lat: Double,
    val lon: Double,
    val heading: Double,
    val altAGL: Double,
    val batteryPct: Double,
    val speedKmh: Double,
    val uin: String,
    val ts: Long
)

data class ViolationAlert(
    val type: String,
    val lat: Double,
    val lon: Double,
    val ts: Long,
    val dismissed: Boolean = false
)

class LiveTrackViewModel(
    private val wsClient: JadsWebSocketClient
) : ViewModel() {

    private val _position = MutableStateFlow<DronePosition?>(null)
    val position: StateFlow<DronePosition?> = _position

    private val _violations = MutableStateFlow<List<ViolationAlert>>(emptyList())
    val violations: StateFlow<List<ViolationAlert>> = _violations

    private val _connectionStatus = MutableStateFlow("CONNECTING")
    val connectionStatus: StateFlow<String> = _connectionStatus

    val trackHistory = mutableListOf<DronePosition>()

    init {
        wsClient.connect()
        viewModelScope.launch {
            wsClient.events.collect { event ->
                when (event) {
                    is WsEvent.Connected -> _connectionStatus.value = "LIVE"
                    is WsEvent.Disconnected -> _connectionStatus.value = "RECONNECTING"
                    is WsEvent.TelemetryPoint -> {
                        val p = DronePosition(
                            lat = event.data.lat,
                            lon = event.data.lon,
                            heading = event.data.headingDeg,
                            altAGL = event.data.altAGL,
                            batteryPct = event.data.batteryPct,
                            speedKmh = event.data.speedKmh,
                            uin = event.data.uin,
                            ts = event.data.ts
                        )
                        _position.value = p
                        trackHistory.add(p)
                        if (trackHistory.size > 500) trackHistory.removeAt(0)
                    }
                    is WsEvent.GeofenceViolation -> {
                        val v = ViolationAlert(event.violationType, event.lat, event.lon, event.ts)
                        _violations.value = _violations.value + v
                    }
                    is WsEvent.BatteryCritical -> {
                        val v = ViolationAlert("BATTERY_CRITICAL", event.data.lat, event.data.lon, event.data.ts)
                        _violations.value = _violations.value + v
                    }
                }
            }
        }
    }

    fun dismissViolation(index: Int) {
        _violations.value = _violations.value.mapIndexed { i, v ->
            if (i == index) v.copy(dismissed = true) else v
        }
    }

    override fun onCleared() {
        wsClient.disconnect()
        super.onCleared()
    }
}

File: jads-ios/Sources/JADS/Telemetry/LiveTrackViewModel.swift
// T07 — ViewModel for live drone tracking on iOS
import Foundation
import Combine
import CoreLocation

class LiveTrackViewModel: ObservableObject {
    @Published var dronePosition: CLLocationCoordinate2D?
    @Published var heading: Double = 0
    @Published var altAGL: Double = 0
    @Published var batteryPct: Double = 0
    @Published var speedKmh: Double = 0
    @Published var uin: String = ""
    @Published var connectionStatus: String = "CONNECTING"
    @Published var violations: [ViolationAlertItem] = []
    @Published var trackHistory: [CLLocationCoordinate2D] = []

    private var wsClient: JadsWebSocketClient
    private var cancellables = Set<AnyCancellable>()

    struct ViolationAlertItem: Identifiable {
        let id = UUID()
        let type: String
        let lat: Double
        let lon: Double
        let ts: Int64
        var dismissed: Bool = false
    }

    init(wsClient: JadsWebSocketClient) {
        self.wsClient = wsClient

        wsClient.events.sink { [weak self] event in
            DispatchQueue.main.async {
                self?.handleEvent(event)
            }
        }.store(in: &cancellables)

        wsClient.connect()
    }

    private func handleEvent(_ event: WsEvent) {
        switch event {
        case .connected:
            connectionStatus = "LIVE"
        case .disconnected:
            connectionStatus = "RECONNECTING"
        case .telemetryPoint(let p):
            dronePosition = CLLocationCoordinate2D(latitude: p.lat, longitude: p.lon)
            heading = p.headingDeg
            altAGL = p.altAGL
            batteryPct = p.batteryPct
            speedKmh = p.speedKmh
            uin = p.uin
            trackHistory.append(CLLocationCoordinate2D(latitude: p.lat, longitude: p.lon))
            if trackHistory.count > 500 { trackHistory.removeFirst() }
        case .geofenceViolation(let v):
            violations.append(ViolationAlertItem(type: v.violationType, lat: v.point.lat, lon: v.point.lon, ts: v.ts))
        case .batteryCritical(let p):
            violations.append(ViolationAlertItem(type: "BATTERY_CRITICAL", lat: p.lat, lon: p.lon, ts: p.ts))
        }
    }

    func dismissViolation(_ item: ViolationAlertItem) {
        if let idx = violations.firstIndex(where: { $0.id == item.id }) {
            violations[idx].dismissed = true
        }
    }

    deinit {
        wsClient.disconnect()
    }
}
